File: Day03/Part01/day03.py
input_file = open("input.txt", "r")
goal = int(input_file.read())
input_file.close()

print("Goal:", goal)

# find minimum square size
size = 1
while (size*size < goal):
    size += 2
print("Size Required:", size)

# start at max tile in biggest required ring
current_tile = size*size

# tuple to hold coordinates of exit
exit_loc = (size // 2, size // 2)

# tuple to hold found location
# will stay (0, 0) if goal is 1  
found_loc = (0, 0)

# look in bottom row
for y in range(size - 1, 0, -1):
    #print("Current tile:", current_tile)
    if current_tile == goal:
        found_loc = (size - 1, y)
    current_tile -= 1

# look in left row
for x in range(size - 1, 0, -1):
    if current_tile == goal:
        found_loc = (x, 0)
    current_tile -= 1

# look in top row
for y in range(size - 1):
    if current_tile == goal:
        found_loc = (0, y)
    current_tile -= 1

# look in right row
for x in range(size - 1):
    if current_tile == goal:
        found_loc = (x, size - 1)
    current_tile -= 1


print("Found Location:", found_loc)

# find required movement
v_movement = found_loc[0] - exit_loc[0]
if (v_movement < 0): v_movement = v_movement // -1
h_movement = found_loc[1] - exit_loc[0]
if (h_movement < 0): h_movement = h_movement // -1
print("Manhattan Distance:", v_movement + h_movement)

File: Day03/Part01/day03.java
import java.io.BufferedReader;
import java.io.FileReader;


public class day03 {

    public static void main(String[] args) {

        // get target from input file
        int target = getTarget("input.txt");
        System.out.println("Target: " + String.valueOf(target));

        // find required size
        int size = getSize(target);
        System.out.println("Size: " + String.valueOf(size));

        // get exit location
        int[] exit_location = getExit(size);
        System.out.println("Exit Location: [" + String.valueOf(exit_location[0]) + ", " + String.valueOf(exit_location[0]) + "]");

        // get target location
        int[] target_location = getTarget(size, target);
        System.out.println("Target Location: [" + String.valueOf(target_location[0]) + ", " + String.valueOf(target_location[1]) + "]");

        // get distance to exit
        int exit_distance = getDistance(target_location, exit_location);
        System.out.println("Manhattan Distance: " + String.valueOf(exit_distance));

    }


    public static int getTarget(String filename) {
        int target = -1;
        try {
            BufferedReader br = new BufferedReader(new FileReader(filename));
            target = Integer.parseInt(br.readLine());
            br.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return target;
    }


    public static int getSize(int target) {
        int t = 1;
        while (t*t < target) {
            t += 2;
        }
        return t;
    }


    public static int[] getExit(int size) {
        int[] loc = new int[2];
        loc[0] = loc[1] = size / 2;
        return loc;
    }


    public static int[] getTarget(int size, int target) {
        int current_node = size * size;
        int[] loc = {0, 0};
        // check bottom row
        for (int i = size - 1; i > 0; i--) {
            if (current_node == target) {
                loc[0] = size - 1;
                loc[1] = i;
            }
            current_node--;
        }
        // check left row
        for (int i = size - 1; i > 0; i--) {
            if (current_node == target) {
                loc[0] = i;
                loc[1] = 0;
            }
            current_node--;
        }
        // check top row
        for (int i = 0; i < size - 1; i++) {
            if (current_node == target) {
                loc[0] = 0;
                loc[1] = i;
            }
            current_node--;
        }
        // check right row
        for (int i = 0; i < size - 1; i++) {
            if (current_node == target) {
                loc[0] = i;
                loc[1] = size - 1;
            }
            current_node--;
        }
        return loc;
    }


    public static int getDistance(int[] target_loc, int[] exit_loc) {
        return Math.abs(target_loc[0] - exit_loc[0]) + Math.abs(target_loc[1] - exit_loc[1]);
    }

} 
